socs/agents/hwp_pid/drivers/pid_controller.py

```diff
@@ -0,0 +1,348 @@
+
+        Args:
+            direction (int): 0 for forward and 1 for backwards
+
+        """
+        if direction == '0':
+            if self.verb:
+                print('Forward')
+            resp = self.send_message("*W02400000")
+            self.direction = 0
+        elif direction == '1':
+            if self.verb:
+                print('Reverse')
+            resp = self.send_message("*W02401388")
+            self.direction = 1
+
+        if self.verb:
+            print(self.return_messages([resp])[0])
+
+    def declare_freq(self, freq):
+        """Declare to memory what the CHWP frequency should be.
+
+        **Note:** Does not actually change the frequency.
+
+        Args:
+            freq (float): Set freqency in Hz. Must be less than 3.5.
+
+        """
+        if float(freq) <= 3.5:
+            self.hex_freq = '0' + self._convert_to_hex(freq, 3)
+            if self.verb:
+                print('Frequency Setpoint = ' + str(freq) + ' Hz')
+        else:
+            if self.verb:
+                print('Invalid Frequency')
+
+    def tune_stop(self):
+        """Set the setpoint to 0 Hz and stop the CHWP."""
+        if self.verb:
+            print('Starting Stop')
+
+        responses = []
+        responses.append(self.send_message("*W0C83"))
+        responses.append(self.send_message("*W01400000"))
+        responses.append(self.send_message("*R01"))
+        responses.append(self.send_message("*Z02"))
+        if self.verb:
+            print(responses)
+            print(self.return_messages(responses))
+
+        stop_params = [0.2, 0, 0]
+        self.set_pid(stop_params)
+
+    def tune_freq(self):
+        """Set the setpoint to currently declared frequency.
+
+        To declare the frequency, use ``PID.declare_freq()``.
+
+        """
+        if self.verb:
+            print('Starting Tune')
+
+        responses = []
+        responses.append(self.send_message("*W0C81"))
+        responses.append(self.send_message(f"*W014{self.hex_freq}"))
+        responses.append(self.send_message("*R01"))
+        responses.append(self.send_message("*Z02"))
+        if self.verb:
+            print(responses)
+            print(self.return_messages(responses))
+
+        tune_params = [0.2, 63, 0]
+        self.set_pid(tune_params)
+
+    def get_freq(self):
+        """Returns the current frequency of the CHWP."""
+        if self.verb:
+            print('Finding CHWP Frequency')
+
+        responses = []
+        responses.append(self.send_message("*X01"))
+        if self.verb:
+            print(responses)
+
+        freq = self.return_messages(responses)[0]
+        return freq
+
+    def get_target(self):
+        """Returns the target frequency of the CHWP."""
+        if self.verb:
+            print('Finding target CHWP Frequency')
+
+        responses = []
+        responses.append(self.send_message("*R01"))
+        if self.verb:
+            print(responses)
+
+        return self.target
```

Comment:
The get_target() function returns the value of self.target. However the value of self.target is not changed within the function itself which could lead to a situation where the value is inaccurate. From what I can tell, self.target only changes when the function return_messages() reads a responce which starts with "R01". This only occurs when self.verb=True and either tune_freq() or tune_stop() is run. If self.verb=False, then self.target will remain 0